### ISSUE TITLE
Unbound global variables

### ISSUE THREAD

**Issue by markuspf** [gap-system/gap#811]
In a freshly started GAP session I get the following junk if I list unbound global variables:

```
gap> Filtered(NamesGVars(), x -> not IsBoundGlobal(x));  
[ "A", "AA", "ATLASREP_TOCFILE", "B", "BINDKEYSTOGAPHANDLER", 
  "BindKeysToGAPHandler", "C", "CustomView", "D", "EXP10_MACFLOAT", 
  "EXP2_MACFLOAT", "EXPM1_MACFLOAT", "EndLineHook", "F", "FG", 
  "FIRST_VIRTUAL_TNUM", "FLOAT_LITERAL_CACHE", "Fam", "G", "H", 
  "HELPBOOKINFOSIXTMP", "I", "IO_FuncToUnpickle", "IO_PkgThingsToRead", 
  "InstallReadlineMacro", "InvocationReadlineMacro", 
  "IsomorphismTypeInfoFiniteSimpleGroup_fun", "KG", "L", "LOG10_MACFLOAT", 
  "LOG1P_MACFLOAT", "LOG2_MACFLOAT", "M", "MB", "MPI_Initialized", "N", 
  "NameGVar", "OnCharReadHookActive", "P", "PrintPromptHook", "R", 
  "READLINEINITLINE", "READ_SMALL_LIB", "RM", "ReadlineInitLine", "S", 
  "STEVES_TRACING", "SavedFloatIndex", "T", "U", "V", "_EVALSTRINGTMP", 
  "_ViewStringForSemigroups", "_ViewStringForSemigroupsGroups", "a", 
  "a_i", "alpha", "arg", "attr", "aut", "b", "bi", "bit", "bk", "bl", 
  "blockmat", "bm", "bn", "c", "ch", "char", "chi", "cl", "cl1", "cl2", 
  "col", "coll", "color", "comp", "cyc", "d", "delta", "dir", "e", "edge", 
  "el", "elem", "elm", "elmlist", "elt", "emb", "empty", "emptyvec", 
  "enum", "esp", "eval", "exp", "ext", "f", "faktor", "fam", "ff", "ffe", 
  "ffelist", "ffus", "fgen", "field", "fnam", "fun", "funclist", "fus", 
  "g", "gen", "gens", "grp", "h", "hash", "hj", "hom", "i", "ic", "id", 
  "idenum", "ii", "ij", "im", "img", "ind", "info", "inn", "inv", 
  "iostream", "it", "iter", "j", "jj", "k", "kbrws", "ker", "key", "l", 
  "l1", "l2", "lambda", "less", "lev", "line", "list", "ll", "lst", "m", 
  "map", "mappinglist", "mat", "mat1", "mat2", "matrix", "mats", 
  "mgmringelms", "modtbl", "msg", "n", "n_blocks", "n_gens", "n_i", "nam", 
  "name", "names", "null", "num", "nz", "o", "obj", "op", "ophom", "opr", 
  "orb", "orbit", "ordtbl", "p", "pair", "paras", "path", "paths", "pcgs", 
  "pcp", "pe", "perm", "person", "phi", "pi", "pnt", "pol", "pol2", "pos", 
  "pre", "prgstring", "prj", "prm", "psi", "q", "quat", "r", "rcl", 
  "record", "rel", "rep", "res", "ri", "rj", "root", "row", "s", "senum", 
  "state", "str", "stream", "string", "sub", "subset", "sym", "t", "t2", 
  "tbl", "tmpclh", "toc", "tom", "triple", "ttbl", "tuple", "type", "u", 
  "unioncl", "unknown", "v", "val", "vals", "value", "values", "vec", "w", 
  "wit", "ww", "x", "xorb", "xset", "y", "z", "zero", "zups", "~" ]
```

Some of these seem to have leaked from somewhere (is this the lambda issue?), others seem to be declared but not bound (yet?). Maybe we can at some point have a look and a cleanup.

**Comment by alex-konovalov** [comment]
That's what I can observe:
```
gap> abc:=42;
42
gap> IsBoundGlobal("abc");
true
gap> "abc" in NamesGVars();
true
gap> UnbindGlobal("abc");
gap> IsBoundGlobal("abc");
false
gap> "abc" in NamesGVars();
true
```
I as surprised at first, but this is documented behaviour: "This function returns an immutable (see 12.6) sorted (see 21.19) list of all the global variable names known to the system. This includes names of variables which were bound but have now been unbound and some other names which have never been bound but have become known to the system by various routes."

**Comment by ChrisJefferson** [comment]
As you say, one issue is that the lambda function ``` zzzz -> zzzz ``` introduces an unbound global variable ```zzzz```

**Comment by ChrisJefferson** [comment]
After #812, now a shorter list:

```
gap> Filtered(NamesGVars(), x -> not IsBoundGlobal(x));
[ "BINDKEYSTOGAPHANDLER", "BindKeysToGAPHandler", "CustomView", "EXP10_MACFLOAT", "F", "FIRST_VIRTUAL_TNUM",
  "FLOAT_LITERAL_CACHE", "HELPBOOKINFOSIXTMP", "IO_FuncToUnpickle", "IO_PkgThingsToRead",
  "InstallReadlineMacro", "InvocationReadlineMacro", "IsomorphismTypeInfoFiniteSimpleGroup_fun",
  "MPI_Initialized", "OnCharReadHookActive", "READLINEINITLINE", "READ_SMALL_LIB", "ReadlineInitLine",
  "RecogsFunnyNameFormatterFunction", "RecogsFunnyWWWURLFunction", "STEVES_TRACING", "SavedFloatIndex",
  "T", "_EVALSTRINGTMP", "_IO_Defines_ChangeDirectoryCurrent", "_ViewStringForSemigroups",
  "_ViewStringForSemigroupsGroups", "a", "dir", "fnam", "gens", "i", "id", "last2", "last3", "mat", "mat1",
  "mat2", "mats", "n_blocks", "n_gens", "op", "p", "pol", "pol2", "time", "tmpclh", "x", "~" ]
```

**Comment by alex-konovalov** [comment]
@markuspf beware of the following (was discussed in gap-dev list in Oct 2010):
```
gap> E;
<Operation "E">
gap> X;
<Operation "Indeterminate">
gap> Z;
function( q ) ... end
```
Now call
```
gap> Filtered(NamesSystemGVars(),name->IsSubset(LETTERS,name));;
```
and then you will have
```
gap> E;
Error, Variable: 'E' must have a value
not in any function at line 8 of *stdin*
gap> X;
Error, Variable: 'X' must have a value
not in any function at line 9 of *stdin*
gap> Z;
Error, Variable: 'Z' must have a value
not in any function at line 10 of *stdin*
```

**Comment by ChrisJefferson** [comment]
The problem here is two-fold:

a) `NamesSystemGVars` gives the "true" copy of the string which we use to represent the variable's name.

b) `IS_SUBSET_SET` calls `PLAIN_LIST` on anything it is given, messing up stringreps.

Here is an example:

```
gap> g := Immutable("E");
"E"
gap> IsMutable(g);
false
gap> IsStringRep(g);
true
gap> IsSubset("A", g);
false
gap> IsStringRep(g);
false
gap> IsMutable(g);
true
gap>
```

Actually, other things get broken, including `Sort`, `List` and `First` and `Int`.

We should probably, for the good of GAP try to remove all calls to `PLAIN_LIST`, except where those calls are to lists that we don't care if their type gets mangled. 

For most uses of `PLAIN_LIST`, I think a `COPY_TO_PLAIN_LIST` would have the same effect, which returned the same list if it was already plain, otherwise made a copy, rather than doing an in-place transformation.

**Comment by fingolfin** [comment]
@alex-konovalov @ChrisJefferson Perhaps this last point (about variable names getting "messed up" by seemingly innocent commands) should be an issue of its own?

**Comment by fingolfin** [comment]
Some of the remaining variables, such as `n_blocks` and `mat`, `mat2`, `mat2` come from the polenta package. Perhaps use `gap -A` to perform the test?

I'll make sure those variables are not set anymore by polenta (they seem to be the result of some left-over test code in an incomplete module of polenta...)

**Comment by ChrisJefferson** [comment]
Issue about breaking variables now in #814, patch in #815. Separate issue #816 for IsSubset

**Comment by fingolfin** [comment]
With #812 merged, and starting gap with `-A`, I now get this:
```
gap> 1;; 2;; 3;; # ensure last, last2, last3 and time are bound
gap> Filtered(NamesGVars(), x -> not IsBoundGlobal(x));
[ "CambridgeMaps", "CustomView", "EXP10_MACFLOAT", "FIRST_VIRTUAL_TNUM",
  "FLOAT_LITERAL_CACHE", "HELPBOOKINFOSIXTMP",
  "IsomorphismTypeInfoFiniteSimpleGroup_fun", "MPI_Initialized",
  "OnCharReadHookActive", "PAGER", "PAGER_OPTIONS", "READ_SMALL_LIB",
  "STEVES_TRACING", "SavedFloatIndex", "SingleHTTPRequest", "_EVALSTRINGTMP",
  "_ViewStringForSemigroups", "_ViewStringForSemigroupsGroups", "dir", "fnam",
  "i", "op", "~" ]
```
Much better. 

I just looked at `STEVES_TRACING` which could be fixed trivially by changing `funcs.c` to set the var only if this tracing code is enabled, i.e.
```
    InitCopyGVar("STEVES_TRACING", &STEVES_TRACING);
```
could be changed to
```
#ifdef TRACEFRAMES
    InitCopyGVar("STEVES_TRACING", &STEVES_TRACING);
#endif
```
Or alternatively, just remove this undocumented debugging hack. Or, if somebody deems it useful, keep it but rename it?

**Comment by fingolfin** [comment]
More things: `CustomView` looks like an ancient hack, which should be removed. `READ_SMALL_LIB` could be avoided by using `CallFuncList` suitably. Not sure where dir, fnam, i, op come from.

**Comment by ChrisJefferson** [comment]
dir and fnam could from trans.grp222: `HideGlobalVariables("dir","fnam");`. The `HideGlobalVariables` function causes the names to come into existence.

We could rewrite that code into a tiny function, something like:

```
if not IsBound(TRANSDEGREES) or TRANSDEGREES=0 then
  TRANSDEGREES := (function()
  local dir, fnam, trans;
    trans := 7;
    dir:= DirectoriesLibrary( "trans" );
    fnam:=Filename( dir, Concatenation("trans",String(trans+1),".grp"));
    while fnam<>fail and IsReadableFile(fnam) do
      trans:=trans+1;
      fnam:=Filename( dir, Concatenation("trans",String(trans+1),".grp"));
    od;
    return trans;
  end)();
fi;
```

I lose track, what's the rules with the trans package? Can I just make pull requests?

**Comment by ChrisJefferson** [comment]
`op` is list.gi:514, once again could do a function which is instantly executed.

**Comment by ChrisJefferson** [comment]
`i` is (I think) `grp/imf.grp`

**Comment by fingolfin** [comment]
With PR #1332 we now get this:
```
gap> 1;; 2;; 3;; # ensure last, last2, last3 and time are bound
gap> Filtered(NamesGVars(), x -> not IsBoundGlobal(x));
[ "CambridgeMaps", "EXP10_MACFLOAT", "FLOAT_LITERAL_CACHE",
  "HELPBOOKINFOSIXTMP", "HPCGAP", "IsomorphismTypeInfoFiniteSimpleGroup_fun",
  "MANEXreadobs", "MPI_Initialized", "OnCharReadHookActive", "PAGER",
  "PAGER_OPTIONS", "READ_SMALL_LIB", "SavedFloatIndex", "SingleHTTPRequest",
  "_EVALSTRINGTMP", "_ViewStringForSemigroups",
  "_ViewStringForSemigroupsGroups" ]
```

**Comment by fingolfin** [comment]
I think we should close this. Or else at last sit down and specify what exactly the goal here is... Getting rid of all of these seems like an endless task, because new ones will keep popping up.

**Comment by fingolfin** [comment]
E.g. one possible resolution would be to add a test to our test suite which checks that this list does not change. Then, if somebody gets rid of another of these, they'll have to adjust the test, which is easy enough; but we'll also notice if somebody introduces a new variable in this list.
